### PR TITLE
fix(api): consolidate PATCH restricted-tags validation into shared helper

### LINKED ISSUE #182: Consolidate scope-validation logic across PATCH /api/articles and /api/import routes
Follow-up to #181 (issue #136).

`PATCH /api/articles/[id]` (`src/app/api/articles/[id]/route.ts:176-210`) and `POST /api/import` (`src/app/api/import/route.ts:252-266`) both validate that the caller can assign the requested `restrictedTags`, but each implements the check inline.

`POST /api/articles` (create path) uses the shared helper `resolveRestrictedTagsForCaller` from `src/lib/api/restricted-scopes.ts`. The PATCH path re-implements the same logic — type check, existence check via `restrictedScope.findMany`, and scope-membership check — instead of calling the helper. The import route uses a new sibling helper `resolveImportRestrictedTags` (different contract: no auto-assign on omission) introduced in #181.

### Suggested consolidation

1. Replace the PATCH route's inline validation (lines 176-208) with a call to `resolveRestrictedTagsForCaller` for the create-time semantics, or to a new `resolvePatchRestrictedTags` helper if PATCH needs different semantics (e.g. only validate when the field is present, which `resolveRestrictedTagsForCaller` does not currently support).
2. Ensure the PATCH route's behavior is byte-identical after consolidation by updating the existing test coverage (or adding new tests if none cover the PATCH scope path).
3. Consider extracting the auto-assign toggle into a named option on the existing helper, since the import and create callers now diverge on exactly this behavior.

### Why this matters

- Two independent implementations of an access-control check is a security smell. A future bugfix that touches the helper will not reach the inline copy.
- The PATCH path also lacks a dedicated test for the unauthorized-scope case (issue #181 only added coverage for the import route).

### Acceptance criteria

- [ ] PATCH `/api/articles/:id` uses the shared helper instead of inline validation
- [ ] Existing PATCH tests still pass
- [ ] New test covers the case where a non-admin key tries to set `restrictedTags` containing a scope it does not have
- [ ] No behavior change visible to clients

### PR DESCRIPTION
**Pull request by SweetSophia**
## Summary

- move PATCH `restrictedTags` validation out of the route and into `resolvePatchRestrictedTags`
- share the scope-membership authorization check with create/import paths
- preserve the existing PATCH API contract byte-for-byte
- add focused unit and route-level regression coverage

## Preserved PATCH behavior

- omitted `restrictedTags` leaves the existing classification unchanged
- `null` and non-array values return HTTP 400
- explicit `[]` declassifies without auto-inheriting caller scopes
- whitespace and duplicate values are not silently normalized
- unknown-tag errors retain the legacy response text
- unauthorized scope assignment returns HTTP 403 without mutating the article

## Verification

- `node --import tsx --test src/__tests__/api/restricted-scopes.test.ts` (42/42)
- `node --import tsx --test src/__tests__/api/articles-patch-restricted-tags.test.ts` (5/5)
- `node --import tsx --test src/__tests__/api/import-route-validation.test.ts` (10/10)
- `npm run typecheck`
- focused ESLint
- `git diff --check`

Closes #182

This pull request consolidates the validation logic for `restrictedTags` on the `PATCH /api/articles/[id]` endpoint into a shared helper function, ensuring consistent authorization checks across article creation, import, and update operations.

Key changes include:

*   **Consolidated Authorization**: Extracted the logic for checking if a caller is authorized to assign specific restricted tags into a new `authorizeRestrictedTags` helper, which is now used by both `normalizeRestrictedTagsForCaller` (for create/import) and `resolvePatchRestrictedTags` (for PATCH).
*   **Stricter PATCH Input Validation**:
    *   `PATCH /api/articles/[id]` now explicitly rejects `null` or `undefined` values for `restrictedTags` with a `400 Bad Request` error, instead of implicitly declassifying the article. Declassification still occurs when an empty array (`[]`) is provided.
    *   The validation ensures that `restrictedTags` is an array containing only non-empty strings.
    *   Tags with leading or trailing whitespace are no longer implicitly trimmed and will be considered invalid if they do not exactly match an existing tag.
*   **Preservation of Duplicate Tags**: The `PATCH` endpoint will now preserve duplicate tags in the `restrictedTags` array as provided by the caller, rather than silently deduplicating them.
*   **Enhanced Error Messages**: Error messages for unknown restricted tags now include a list of valid tags, providing more informative feedback to API consumers.

### CODE REVIEW

File: src/lib/api/restricted-scopes.ts [hunk 1]
```diff
@@ -97,6 +97,35 @@ export async function resolveImportRestrictedTags(
...
+  lookup: RestrictedScopeLookup = findExistingRestrictedScopes,
+): Promise<RestrictedTagsResult> {
+  if (value === undefined || value === null) {
+    return { ok: true, value: [] };
+  }
+  if (Array.isArray(value) && value.length === 0) {
+    return { ok: true, value: [] };
+  }
+  return resolveRestrictedTagsForCaller(value, allowedScopes, lookup);
+}
```
[the remark this answers is not in the document]
Comment (SweetSophia) [reply src/lib/api/restricted-scopes.ts:149]: Addressed in 2377d07 with a different shape. PATCH cannot delegate to the import helper because issue #182 requires no client-visible behavior change: PATCH rejects null, preserves duplicates/whitespace, and retains its legacy unknown-tag error. The sensitive membership check is now shared through `authorizeRestrictedTags`.

File: src/lib/api/restricted-scopes.ts [hunk 2]
```diff
@@ -97,6 +97,35 @@ export async function resolveImportRestrictedTags(
...
+ * existence-checked, and scope-membership-checked — all via the shared
+ * `resolveRestrictedTagsForCaller` core so a fix to the access-control logic
+ * reaches the create, import, and PATCH paths together (see issue #182).
+ *
+ * This is intentionally a separate, PATCH-named entry point rather than a flag
+ * on the create helper: the create path auto-assigns caller scopes on omission,
+ * whereas PATCH must not. Keeping the contracts in named functions makes the
+ * divergence visible at each call site.
+ */
+export async function resolvePatchRestrictedTags(
```
[the remark this answers is not in the document]
Comment (SweetSophia) [reply src/lib/api/restricted-scopes.ts:124]: Addressed in 2377d07. PATCH and import are intentionally not identical once the issue's byte-identical compatibility requirement is enforced. The duplicated security-sensitive membership logic was extracted into `authorizeRestrictedTags`; PATCH keeps its legacy validation and response contract.

File: src/__tests__/api/articles-patch-restricted-tags.test.ts [hunk 3]
```diff
@@ -0,0 +1,213 @@
...
+    assert.equal(response.status, 200, "authorized PATCH should succeed");
+    assert.deepEqual(
+      body.restrictedTags,
+      [TEST_SCOPE_OWNED],
+      "restrictedTags should be persisted",
+    );
+  } finally {
+    await cleanupSharedFixtures(prisma);
+  }
+});
```
[the remark this answers is not in the document]
Comment (SweetSophia) [reply src/__tests__/api/articles-patch-restricted-tags.test.ts:191]: Implemented in 2377d07. The authorized PATCH test now reloads the article through Prisma and asserts the stored `restrictedTags` value.

File: src/__tests__/api/articles-patch-restricted-tags.test.ts [hunk 4]
```diff
@@ -0,0 +1,213 @@
...
+import crypto from "node:crypto";
+import type { NextRequest } from "next/server";
+import type { PrismaClient } from "@prisma/client";
+
+// Route-level test for the PATCH /api/articles/[id] restricted-tags path.
+// Mirrors import-route-validation.test.ts: it needs a reachable DATABASE_URL
+// and the local dev stack, and exercises the real route handler end-to-end.
+if (!process.env.DATABASE_URL) {
+  throw new Error("DATABASE_URL environment variable must be set for tests");
+}
```
[the remark this answers is not in the document]
Comment (SweetSophia) [reply src/__tests__/api/articles-patch-restricted-tags.test.ts:12]: Not applying. This is a test-only prerequisite guard, consistent with the repository's database integration tests; it does not parse application runtime configuration. Adding Zod here would add unrelated machinery without improving the test boundary.

File: src/app/api/articles/[id]/route.ts [hunk 5]
```diff
@@ -173,39 +174,21 @@ export async function PATCH(request: NextRequest, { params }: { params: Promise<
...
+    // restrictedTags — validate via the shared scope helper (issue #182).
+    // PATCH semantics: an explicit [] declassifies and never auto-inherits the
+    // caller's scopes; only present tags are type/existence/membership-checked.
     if (restrictedTags !== undefined) {
-      if (!Array.isArray(restrictedTags) || !(restrictedTags as unknown[]).every((t) => typeof t === "string" && t.length > 0)) {
-        return NextResponse.json({ error: "restrictedTags must be an array of non-empty strings" }, { status: 400 });
+      const restrictedTagsResult = await resolvePatchRestrictedTags(
+        restrictedTags,
+        auth.auth.allowedScopes,
+      );
```
[the remark this answers is not in the document]
Comment (SweetSophia) [reply src/app/api/articles/[id]/route.ts:184]: Not applying. This route already uses field-specific validation to preserve established status codes and error text. `resolvePatchRestrictedTags` now validates the complete PATCH contract, including null, element types, existence, and caller authorization. A route-wide Zod migration is outside issue #182 and would risk client-visible changes.

File: src/lib/api/restricted-scopes.ts [hunk 6]
```diff
@@ -97,19 +112,58 @@ export async function resolveImportRestrictedTags(
...
+      status: 400,
+      error: "restrictedTags must be an array of non-empty strings",
+    };
+  }
+
+  const existing = await validateRestrictedTagsExist(
+    value,
+    lookup,
+    (invalidTags, validTags) =>
+      `Unknown restricted tag(s): ${invalidTags.join(", ")}. Valid tags: ${validTags.join(", ")}`,
```
[the remark this answers is not in the document]
Comment (SweetSophia) [reply src/lib/api/restricted-scopes.ts:144]: Not applying. The premise is stale: the PR description now explicitly says the legacy PATCH error text is preserved, matching issue #182's no-client-visible-change acceptance criterion. Also, the old query selected only valid tags from the caller's submitted list, not the full restricted-scope catalog, so this is not the claimed catalog disclosure. Removing the suffix belongs in a separately specified API-breaking/security-hardening change.

File: src/lib/api/restricted-scopes.ts [hunk 7]
```diff
@@ -97,19 +112,58 @@ export async function resolveImportRestrictedTags(
...
+  if (
+    !Array.isArray(value) ||
+    !value.every((tag) => typeof tag === "string" && tag.length > 0)
+  ) {
+    return {
+      ok: false,
+      status: 400,
+      error: "restrictedTags must be an array of non-empty strings",
+    };
+  }
```
[the remark this answers is not in the document]
Comment (SweetSophia) [reply src/lib/api/restricted-scopes.ts:138]: Not applying. This is a style-rule preference, not a correctness defect. The repository does not declare Zod as a direct dependency, and this helper deliberately preserves the existing field-specific validation, status code, and error text required by issue #182. Adding a new validation dependency or changing the validation path would expand scope without improving this boundary.
